### PR TITLE
[SPARK-9750][MLlib] Improve equals on SparseMatrix and DenseMatrix

### PR DESCRIPTION
**Pull request by feynmanliang**
Adds unit test for `equals` on `mllib.linalg.Matrix` class and `equals` to both `SparseMatrix` and `DenseMatrix`. Supports equality testing between `SparseMatrix` and `DenseMatrix`.

@mengxr

### CODE REVIEW

**Comment by feynmanliang** [comment]
To start the discussion; should a `SparseMatrix` and a `DenseMatrix` be equal if they contain the same underlying content (i.e. do we define `==` as equality in semantics or as equality in representation)?

**Comment by feynmanliang** [comment]
Jenkins test please

**Comment by feynmanliang** [comment]
Jenkins test this please

**Comment by jkbradley** [comment]
> To start the discussion; should a SparseMatrix and a DenseMatrix be equal if they contain the same underlying content (i.e. do we define == as equality in semantics or as equality in representation)?

If we follow Breeze's example, then we should define equality in semantics, not representation.  I think this PR is valid.

**Comment by jkbradley** [comment]
(From JIRA): This JIRA should probably be expanded to include DenseMatrix too. Currently, DenseMatrix.equals always returns false for comparisons with SparseMatrix instances.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 1]
```diff
@@ -519,6 +519,11 @@ class SparseMatrix(
       rowIndices: Array[Int],
       values: Array[Double]) = this(numRows, numCols, colPtrs, rowIndices, values, false)
 
+  override def equals(o: Any): Boolean = o match {
+    case m: SparseMatrix => toBreeze == m.toBreeze
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: This case should be for Matrix, not just SparseMatrix.

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala [hunk 2]
```diff
@@ -74,6 +74,22 @@ class MatricesSuite extends SparkFunSuite {
     }
   }
 
+  test("equals") {
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala]: Could you expand this to compare dense with sparse?

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala (hunk 1 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: OK.

Will also change `DenseMatrix.equals`.

File: mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala (hunk 2 above)
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/linalg/MatricesSuite.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 3]
```diff
@@ -519,6 +519,12 @@ class SparseMatrix(
       rowIndices: Array[Int],
       values: Array[Double]) = this(numRows, numCols, colPtrs, rowIndices, values, false)
 
+  override def equals(o: Any): Boolean = o match {
+    case m: Matrix =>
+      m.numRows == numRows && m.numCols == numCols && Arrays.equals(toArray, m.toArray)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: I like what you had before better.  Using Breeze, I think you'd avoid converting to dense representations.  With toArray, it becomes dense.

Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Totally forgot about that, thanks.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala [hunk 4]
```diff
@@ -257,7 +257,7 @@ class DenseMatrix(
     this(numRows, numCols, values, false)
 
   override def equals(o: Any): Boolean = o match {
-    case m: DenseMatrix =>
+    case m: Matrix =>
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: You could replace toArray with toBreeze here too, in case "o" is sparse

Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala]: Done

**Comment by jkbradley** [comment]
LGTM pending tests

**Comment by jkbradley** [comment]
Merging with master

**Comment by mengxr** [comment]
@feynmanliang The contract for Java's `Object` is that `a.equals(b)` implies `a.hashCode == b.hashCode`. So usually we need to implement both. The problem with `hashCode` is that we shouldn't compute it based on all values, which could be very expensive. You can use the implementation of `Vector.hashCode` as a template, but that requires some changes to avoid hash code collisions.

**Comment by feynmanliang** [comment]
@mengxr made SPARK-9919 to track this

**Pull request closed without merging**